tsfresh/utilities/dataframe_functions.py

```diff
@@ -372,7 +375,55 @@ def _normalize_input_to_internal_representation(timeseries_container, column_id,
     return timeseries_container, column_id, column_kind, column_value
 
 
-def roll_time_series(df_or_dict, column_id, column_sort, column_kind, rolling_direction, max_timeshift=None):
+def _roll_out_time_series(time_shift, grouped_data, rolling_direction, max_timeshift, min_timeshift,
+                          column_sort, column_id):
+    """
+    Internal helper function for roll_time_series.
+    This function has the task to extract the rolled forecast data frame of the number `time_shift`.
+    This means it is `time_shift` in the future - starting counting from the first row
+    for each id (and kind).
+    This means we cut out the data until `time_shift`.
+    The first row we cut out is either 0 or given by the maximal allowed length of `max_timeshift`.
+    for a negative rolling direction it is reversed
+    """
+    def _f(x):
+        if rolling_direction > 0:
+            shift_until = time_shift
+            shift_from = max(shift_until - max_timeshift - 1, 0)
+
+            df_temp = x.iloc[shift_from:shift_until] if shift_until <= len(x) else None
+        else:
+            shift_from = max(time_shift - 1, 0)
+            shift_until = shift_from + max_timeshift + 1
+
+            df_temp = x.iloc[shift_from:shift_until]
+
+        if df_temp is None or len(df_temp) < min_timeshift + 1:
```

Comment:
is +1 really needed?

Reply:
The way I see it: If you request a minimal time shift of 3, you want to have all the data which includes at least 3 time steps in the past.
In our nomenclature, that would be "t, t-1, t-2, t-3", so this would give us a length of 4.